runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/WindowingStrategyTranslation.java

```diff
@@ -295,10 +299,17 @@ public static FunctionSpec toProto(WindowFn<?, ?> windowFn, SdkComponents compon
             .setAllowedLateness(windowingStrategy.getAllowedLateness().getMillis())
             .setTrigger(TriggerTranslation.toProto(windowingStrategy.getTrigger()))
             .setWindowFn(windowFnSpec)
-            .setAssignsToOneWindow(windowingStrategy.getWindowFn().assignsToOneWindow())
+            .setAssignsToOneWindow(windowFn.assignsToOneWindow())
+            .setMergeStatus(
+                windowFn instanceof InvalidWindows
+                    ? (((InvalidWindows) windowFn).getOriginalWindowFn().isNonMerging()
```

Comment:
probably stupid question: if the InvalidWindow only happens after GBK or combine, does the original window fn matter? should the MergeStatus always be ALREADY_MERGED for InvalidWindow? How should runner tread ALREADY_MERGED differently than NON_MERGING?

Reply:
It is a good question. I think the original WindowFn is just for giving a good error message. It is also used to check compatibility of `Flatten` of two PCollections.

Reply:
`InvalidWindows` is only used in exactly one place, only for merging `originalWindowFn`: https://github.com/apache/beam/blob/7c43ab6a8df9b23caa7321fddff9a032a71908f6/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupByKey.java#L216

Reply:
Are we assuming that usage here (not just the InvalidWindows spec) ? If so probably this at least needs a comment about that assumption.

Reply:
Per dev@ list discussion, I'm going to change this to not special-case `InvalidWindows`. For now: if it is a merging WindowFn it says NEEDS_MERGE, otherwise it doesn't.

Once I finish adjusting the SDKs to actually track ALREADY_MERGED bit, I will adjust the translation too.